[ph system] Various properties defined in NM lua vs in each PH lua

MowFord opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## Preliminary Info

I didn't see any issues related to ongoing work with the PH system, so hopefully this is good to introduce.

I found some interesting behavior of the `xi.mob.phOnDespawn` function
- `phList` does need to be a table that keys `phId` -> `nmId`
  - a couple spots utilize a dummy entry in the `phList` to block the chance of an NM spawn from a PH if another mob/nm is up
  - quite a few NMs have multiple entity IDs, where if one is chosen to spawn, the other needs to be blocked
- some PH lua files defined the full list of PHs for an NM, and some only defined the ones related to that particular PH-mob lua name
  - The splitting of `phList` from the IDs.lua into mob lua files was a good one to make a clean shift-proof system, but it came with the possibility of doing this and/or duplicating data
- same with `spawnList`. It's good to move that out of sql, but I propose we define it on the NM and not (potentially duplicated) in each PH lua file
  - It also provides a consistent way to migrate other NMs with calls to `UpdateNMSpawnPoint` in their despawn function. We can move the spawnList to the entity and change the call to be something like `xi.mob.updateNMSpawnPoint(mob:getName())` and do a similar extraction from the global table

## What does this pull request do?

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->

This PR seeks to consolidate information about an NM to its lua file. It came about when digging around trying to find out which PHs spawn Poisonhand Gnadgad
- changes the behavior of `xi.mob.onPhDespawn` slightly
  - makes `params.spawnList` still an option to override the default, but the default being `spawnList` defined on the NM directly
  - adds `params.doNotEnablePhSpawn` to be used when PHs are scripted spawns themselves, so we don't enable respawning when NM dies
  - renames the `phList` parameter to `nmPhId`
    - will still accept the old method for now
    - if `nmPhId` is a number, instead of the current method of being a table, the NM's entity object is pulled from the global table, and the `phList` is extracted from that (central) location
  - I've also put more guard clauses to reduce the indention of the function
    - which unfortunately makes the diff very hard to read, i recommend simply looking at the new function's flow.
    - most sections weren't edited, just shift+tabbed because of not being within an if-block anymore

I believe these changes to simplify the definition of PH-style NMs, and reduce duplication. 

I went ahead and converted a selection of PH-Style NMs to ensure it works for all the types of spawn conditions

Edit: my changes bumped cyclomatic complexity past the limit, so i crunched the boolean param enforcement into a table/for loop

Edit2: After talking with Xaver, I've updated some things above. Also
- `xi.mob.updateNMSpawnPoint` to be in its final form:
  - accepts mob as an id or an entity
  - accepts an optional `spawnPoints` table
  - if `spawnPoints` isn't found as a param or attached to the mob's lua object, then it calls the legacy function `UpdateNMSpawnPoint`, and displays no message if successful
    - All calls to `UpdateNMSpawnPoint` can be string replaced directly
    - as we retire sql data, we would move the spawn points to `entity.spawnPoints` tables in the NM's lua
    - `Zone.lua` calls would access the table just as easily as calls within the mob's `onMobDespawn` function

## Steps to test these changes

<!-- Clear and detailed steps to test your changes here -->
- for each PH lua, it greatly simplifies the process if you 
  - add to the params: `params.immediate = true`
  - set chance to `100`
  - set cooldown to `1`
  - (optional): update the `spawnPoints` table in the NM's lua to see that it's doing a lottery on this list to update the NM spawn point each time it's selected to spawn
- Poisonhand
  - `!zone 149`
  - `!gotoname Poisonhand_Gnadgad`
  - note the ID and offset it to find one of the PHs
  - go to one of the PHs and kill it
  - see that NM spawns like usual
- Swamfisk
  - `!zone 101`
  - same as above to get NM id
  - See that the proper NM spawns based on which you kill
- Rampaging Ram/Steelfleece
  - This was a fun one.
    - this is a good example of how defining things _on the NM's lua_ helps organization
    - It is actually not correct in current LSB due to:
      - tremor ram being defined as a PH for steelfleece, which is [incorrect](https://www.bg-wiki.com/ffxi/Steelfleece_Baldarich)
      -  `DESPAWN` listeners happening _after_ `onMobDespawn`, causing `xi.mob.phOnDespawn` being blocked if your own ID is in the value side of the `phList`
        - That caused an issue where steelfleece blocks rampaging and rampaging blocks steelfleece, but is actually of no consequence due to the fact that there's only one rampaging
          - i.e. if rampaging is up, killing tremor rams will not spawn another rampaging
          - if steelfleece is up, rampaging _won't_ be up
          - if steelfleece is up, rampaging will be blocked appropriately with the updates in this PR
  - to test:
    - `!zone 108`
    - the below tests assume you have done the recommendation of making chance 100% for both tiers
    - `!gotoname Tremor_Ram 1` or 2
    - kill it
    - `!gotoname Rampaging_Ram`
    - note the ID for later
    - kill it
    - `!gotoname Steelfleece_Baldarich`
    - see that it's up
    - go back to other tremor ram, kill it, and see that it doesn't pop Rampaging
    - kill Steelfleece
    - see that Rampaging respawn time is still zero (it doesn't get enabled incorrectly)
      - `!exec print(GetMobByID(17219886):getRespawnTime())
- Finally, to test the updated `xi.mob.updateNMSpawnPoint`, you can simply use `!exec`
  - test mobid passing: `!exec xi.mob.updateNMSpawnPoint(target:getID())`
  - test mob entity passing: `!exec xi.mob.updateNMSpawnPoint(target)`
  - and optionally pass a spawnPointsOverride: `!exec xi.mob.updateNMSpawnPoint(target, {{x=1,y=1,z=1}})`
  - then either way you did it, `!exec print(target:getSpawnPos())` to see that it's updated as expected

Thanks for coming to my TED talk

MowFord committed:
Update phOnDespawn to expect things defined on the mob

MowFord committed:
Convert Poisonhand Gnadgad to centralized phList

MowFord committed:
Convert Swamfisk to centralized phList

MowFord committed:
Convert Rampaging and Steelfleece PH system

- to centralized phList
- fix steelfleece being tied to tremor ram
- fix steelfleece spawning from rampaging
- fix rampaging to not respawn after killing steelfleece

MowFord committed:
Add definitions to TMobEntity for phList and spawnPoints

Xaver-DaRed commented on scripts/zones/Davoi/mobs/Orcish_Beastrider.lua:
```diff
@@ -3,22 +3,14 @@
 
 entity.onMobDespawn = function(mob)
-    xi.mob.phOnDespawn(mob, poisonhandPHTable, 10, 3600) -- 1 hour
+    xi.mob.phOnDespawn(mob, 'Poisonhand_Gnadgad', 10, 3600) -- 1 hour
```
Just in general, I would rather we passed the mob ID instead of the name as an argument

MowFord replied:
That's of course up to you, but the name is literally what we're mapping to, as some of these nms have multiple IDs (though they should all have the same name

I.e. for swamfisk we would pass `ID.mob.SWAMFISK[1]` then rearrange the nmId logic in the 'phOnDespawn' function

As far as I know, lua names don't change from dat exports so it should be the same, and makes the logic easier on the other side, but it's possible either way

MowFord replied:
Updated to take nm ID as the second parameter of `xi.mob.phOnDespawn`, and added more guard clauses since we get the variables earlier in the flow

Xaver-DaRed commented on scripts/zones/Davoi/mobs/Poisonhand_Gnadgad.lua:
```diff
@@ -2,11 +2,29 @@
+{
+    { x = -61.045, y = -0.517, z = 41.996 },
+}
+
```
If they dont need to be global, leave them local, please

MowFord replied:
These values are being grabbed by the pH function, so they have to be global

MowFord replied:
After discussing in DMs, I've updated the `xi.mob.updateNMSpawnPoint` function to be more generalized so it can be used for any mob. It's reverse compatible with how the `local spawnPoints` tables were being used, but it also is a drop-in replacement for `UpdateNMSpawnPoint` by accepting a mob id or mob entity object.

MowFord committed:
mobs.lua function updates

- xi.mob.phOnDespawn finds nm based on explicit ID as param
- xi.mob.updateNMSpawnPoint is now generalized enough to be used anywhere

MowFord committed:
Update padfoot to use the new spawnPoints paradigm

WinterSolstice8 commented on scripts/globals/mobs.lua:
```diff
@@ -31,46 +31,111 @@ local function lotteryPrimed(phList)
+-- This function is used to replace UpdateNMSpawnPoints() inside the Zone.lua files and the NM despawn scripts
+-- - mobParam can either be a mobid or a mob entity object
+-- it either accepts a table of spawn points to randomize, or looks to the mob's cached lua object for a spawnPoints entry
+xi.mob.updateNMSpawnPoint = function(mobParam, spawnPointsOverride)
```
I think this is complicated enough to warrant LLS hinting now

I think it's this (don't quote me on this)
```lua
---@param mobParam integer|CBaseEntity
---@param spawnPointsOverride table?
xi.mob.updateNMSpawnPoint = function(mobParam, spawnPointsOverride)
```

WinterSolstice8 replied:
Updated, looks like mobParam isn't actually optional

MowFord replied:
Sounds good, let's see what LLS thinks about the changes

MowFord replied:
Aiight round 2. Tested locally that ph and spawnPoint functions still work, just added some extra type checking to hopefully make LLS happy

WinterSolstice8 commented on scripts/globals/mobs.lua:
```diff
@@ -31,46 +31,111 @@ local function lotteryPrimed(phList)
 
 -- potential lottery placeholder was killed
-xi.mob.phOnDespawn = function(ph, phList, chance, cooldown, params)
+xi.mob.phOnDespawn = function(ph, phNmId, chance, cooldown, params)
```
And for this one I think it would be...
```lua
---@param ph CBaseEntity
---@param phNmId integer
---@param chance integer
---@param cooldown integer
---@param params table?
xi.mob.phOnDespawn = function(ph, phNmId, chance, cooldown, params)
```

WinterSolstice8 reviewed (approved):
This must have been a lot of effort, good work
